highfidelity/hifi
# Case 20080: v3 + v4 procedurals and misc shader fixes

https://highfidelity.manuscript.com/f/cases/20080/Procedural-Shader-bugs-improvements

Test plan:
- Perform these tests on various graphics cards: NVIDIA, AMD, test on Mac
- Run [this](https://gist.githubusercontent.com/SamGondelman/d2fe968f275be983ba2ed328437f1add/raw/17a96571005d4b0a647a44b0c1eddfc28ca4acad/CreateProcedurals.js).  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191234-b6738480-f32a-11e8-9617-7dfceb5bd0d2.png)
- Press "u".  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191249-c68b6400-f32a-11e8-9bab-d8ebbfaa4c82.png)
- Press "u" again.  Press "i".  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191286-dc992480-f32a-11e8-803b-145216eb3b16.png)
- Press "i" again.  Press "h".  Open luci.js and select the "Roughness" debug framebuffer.  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191576-eb340b80-f32b-11e8-9cd3-ad630ef9a6b5.png)
- Press "h" again.  Press "j".  Select the "Metallic" debug framebuffer.  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191612-0737ad00-f32c-11e8-94e9-45f3b62fccf2.png)
- Press "j" again.  Press "k".  Select the "Occlussion" debug framebuffer.  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191639-29312f80-f32c-11e8-8fba-615950335856.png)
- Press "k" again.  Press "l".  Select the "Scattering" debug framebuffer.  You'll see this:
![image](https://user-images.githubusercontent.com/7650116/49191655-3d752c80-f32c-11e8-8898-47e79736cd19.png)
- Run this from your console:
```
Entities.addEntity({
	type: "Box",
	position: Vec3.sum(MyAvatar.position, Vec3.multiply(4.0, Quat.getForward(MyAvatar.orientation))),
	dimensions: 0.5,
	collisionless: true,
	userData: JSON.stringify({ "ProceduralEntity": {"version":4,"shaderUrl":"https://gist.githubusercontent.com/SamGondelman/932c8c03fd39b0549f0a07c0a736abd9/raw/85e047ff15e68c29c730823421b52a4024311753/raymarch.fs","uniforms":{"scene":0}}})
})
```
- Make sure the following work on Mac and PC.
- You should see some grey spheres moving around a green rippling ball that morphs into a beach ball. They should be lit correctly by the keylight (try changing the keylight. The shadowed areas on the spheres should change). Try placing other lights and entities around. The spheres should be lit by the local lights and should depth-test properly with the objects.
- Look at the balls in VR. They should appear 3D. Pass your hand through them. They should intersect your hand properly (with respect to depth, not physics).
